### PR TITLE
fix(regexp): treat non-quantifier { as Annex B literal

### LINKED ISSUE #359: Regex containing a literal unescaped { or } is rejected instead of treated as an Annex B ExtendedPatternCharacter
## Summary

Per Annex B's `ExtendedPatternCharacter` production (used for non-`u`/non-`v`
mode regexes), a `{` that isn't the start of a valid quantifier (`{n}`,
`{n,}`, `{n,m}`) should be treated as a literal character rather than
rejected. jsse rejects it outright — and worse, a regex *literal* containing
this pattern fails at parse time (uncatchable), while `new RegExp(...)` at
least raises a catchable SyntaxError.

## Repro

```
$ node -e 'console.log(/{}/g.source)'
{}

$ jsse -e 'console.log(/{}/g.source)'
SyntaxError: Invalid regular expression: /{}/ : Nothing to repeat
  (fails to even parse the script — the whole file is rejected, not a
  catchable runtime error)

$ node -e 'console.log(new RegExp("{}", "g").source)'
{}

$ jsse -e 'console.log(new RegExp("{}", "g").source)'
Uncaught SyntaxError: Invalid regular expression: /{}/ : Nothing to repeat
  (catchable this time, but still wrong per spec)
```

## Discovered via

Manual exploration while integrating the esprima Node-compat library harness
(#295). Not currently surfaced by any test262/esprima corpus scenario checked
so far.

## Likely area

`src/interpreter/builtins/regexp.rs` — regex compiler's quantifier/atom
grammar needs the Annex B `ExtendedPatternCharacter` fallback for a lone
`{`/`}` not forming a valid quantifier (non-unicode-mode only).

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- `validate_js_pattern`'s "is this `{` a quantifier" pre-checks accepted any digit/comma run terminated by `}` as a valid quantifier, including empty content (`{}`). Per ECMA-262 Annex B's `InvalidBracedQuantifier`/`ExtendedPatternCharacter` grammar, a brace-quantifier requires at least one leading digit (`{n}`, `{n,}`, `{n,m}`); anything else must fall through to `ExtendedPatternCharacter` and be treated as an ordinary literal character in non-Unicode mode rather than raising a SyntaxError.
- Added a shared `quantifier_brace_end` helper that correctly implements the grammar, and used it to fix the two buggy call sites (the assertion-followed-by-quantifier check, and the before-atom check). The other two call sites already validated digits correctly via `parse::<u64>()`, so they were left unchanged.
- Verified the fix against Node across ~30 brace/quantifier edge cases (both Unicode and non-Unicode modes) and against the actual ECMA-262 Annex B spec grammar text.

Fixes #359

## Test plan

- [x] `cargo test --release` — 340 unit tests + mutation-oracle test, all passing
- [x] `uv run python scripts/run-custom-tests.py` — 9/9 passing
- [x] `./scripts/lint.sh` — rustfmt + clippy clean
- [x] `uv run python scripts/run-test262.py -j 32` — 99,568/99,568 (100.00%), 0 regressions, 323 new passes
- [x] Manual repro confirms the bug is fixed: `jsse -e 'console.log(/{}/g.source)'` and `new RegExp("{}", "g")` now both return `{}` instead of raising a SyntaxError
- [x] New regression test added: `test262-extra/RegExp-annexb-brace-literal.js`

### CODE REVIEW

**Comment by pmatos** [comment]
@claude Regarding your automated review comment (no issues found):

**Not an issue** — the review reported a clean pass ("No issues found. Checked for bugs and CLAUDE.md compliance.") with no concrete finding, code reference, or requested change, so there is nothing to fix or track. No action taken.

_This assessment was made by two independent AI reviewers (Claude Opus and Codex GPT-5.6-sol). If you disagree, please reply and we'll re-evaluate._

File: src/interpreter/builtins/regexp.rs [hunk 1]
```diff
@@ -4510,30 +4525,16 @@ pub(crate) fn validate_js_pattern(source: &str, flags: &str) -> Result<(), Strin
...
-                            ));
-                        }
-                        has_atom = true;
-                        i += 1;
-                        continue;
+                if c == '{' && quantifier_brace_end(&chars, i, len).is_none() {
+                    if unicode {
+                        return Err(format!(
+                            "Invalid regular expression: /{}/ : Lone quantifier brackets",
+                            source
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/interpreter/builtins/regexp.rs:4546]: Fixed in `5d77500`.

You're right that this was still broken, and it turned out to be broader than the lookbehind case you flagged: `translate_js_pattern_ex` (the pattern → fancy-regex translation step, separate from `validate_js_pattern`'s syntax gate) was passing comma-leading brace forms like `{,}`/`{,5}` through unescaped in *every* context where `has_atom` resets — not just after a lookbehind, but also bare, after a plain atom (`a{,}`), after a real quantifier (`a*{,}`), and after a lookahead. fancy-regex is more lenient than ECMAScript here (it accepts `{,n}` as shorthand for `{0,n}`), so these were silently matched as real quantifiers instead of literal text.

Two fixes, both gated on the existing `quantifier_brace_end` helper from the previous commit:
- The generic per-character fallthrough in `translate_js_pattern_ex` now escapes `{` whenever it doesn't open a valid quantifier.
- The Annex B quantified-lookahead special case (which strips/keeps a lookahead assertion based on its trailing quantifier) was misidentifying non-quantifier `{` as a real quantifier trigger and silently dropping the `{` character entirely — that's now gated the same way.

Validation: added `assertExecMatch` coverage in `test262-extra/RegExp-annexb-brace-literal.js` asserting actual matched text (not just `.source`/`.test()`) for comma-leading forms across all the affected contexts, cross-checked against Node. Also re-verified the previously-working no-comma forms (`{}`, `a{}`, `{1,2,3}`) and real quantifiers (`a{2,3}` etc.) are unaffected. Full test262: 99,568/99,568 (100.00%), 0 regressions.

File: src/interpreter/builtins/regexp.rs [hunk 2]
```diff
@@ -3061,6 +3064,17 @@ pub(super) fn translate_js_pattern_ex(
             continue;
         }
 
+        // Annex B: a '{' that doesn't open a valid quantifier is an ordinary
+        // literal character. fancy-regex is more lenient than ECMAScript here
+        // (e.g. it accepts `{,n}` as shorthand for `{0,n}`), so it must be
+        // escaped explicitly rather than passed through, or it and the
+        // characters after it get reinterpreted as a real quantifier.
+        if c == '{' && !in_char_class && quantifier_brace_end(&chars, i, len).is_none() {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/interpreter/builtins/regexp.rs:3072]: Fixed in `69c8fa4`.

Confirmed — `nq_has_min0`'s `{` arm only checked that the character right after `{` was `'0'`, without validating the rest of the form (`{0,foo}` fails ECMAScript's QuantifierPrefix grammar since `foo` isn't `DecimalDigits`, so it's an Annex B literal, not a real `{0,...}` quantifier). Gated it on the existing `quantifier_brace_end` helper, same as the two prior fixes.

While auditing every other brace-quantifier-detection site in the file for the same pattern (rather than waiting for another round), I found the identical bug shape in `build_quantified_parent_map`'s `)`-followed-by-quantifier check: it also treated `{0,foo}` as a real min-zero quantifier, which fed into the stale-capture-clearing pass and wrongly cleared a nested capture that had legitimately matched (`/(?:(?=(.))){0,foo}/.exec("{0,foo}")` returned `["{0,foo}", null]` on jsse instead of `["{0,foo}", "{"]`). Fixed the same way.

I also hit a fourth bug during the audit that's real but unrelated to braces entirely (`/(a*|dc??)*/` matches `""` on jsse vs `"d"` on Node — an alternation/lazy-marker interaction in the same nullable-rewrite pass). That one is out of scope for this PR and tracked separately as #370.

Validation: added `assertExecMatch`/`assertExecGroups` coverage in `test262-extra/RegExp-annexb-brace-literal.js` for both bugs. Full test262: 99,568/99,568 (100.00%), 0 regressions.
